x/incentive/keeper/grpc_query.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	var usdxFactors types.RewardIndexes
+	s.keeper.IterateUSDXMintingRewardFactors(sdkCtx, func(collateralType string, factor sdk.Dec) (stop bool) {
+		usdxFactors = usdxFactors.With(collateralType, factor)
+		return false
+	})
+
+	var supplyFactors types.MultiRewardIndexes
+	s.keeper.IterateHardSupplyRewardIndexes(sdkCtx, func(denom string, indexes types.RewardIndexes) (stop bool) {
+		supplyFactors = supplyFactors.With(denom, indexes)
+		return false
+	})
+
+	var borrowFactors types.MultiRewardIndexes
+	s.keeper.IterateHardBorrowRewardIndexes(sdkCtx, func(denom string, indexes types.RewardIndexes) (stop bool) {
+		borrowFactors = borrowFactors.With(denom, indexes)
+		return false
+	})
+
+	var delegatorFactors types.MultiRewardIndexes
+	s.keeper.IterateDelegatorRewardIndexes(sdkCtx, func(denom string, indexes types.RewardIndexes) (stop bool) {
+		delegatorFactors = delegatorFactors.With(denom, indexes)
+		return false
+	})
+
+	var swapFactors types.MultiRewardIndexes
+	s.keeper.IterateSwapRewardIndexes(sdkCtx, func(poolID string, indexes types.RewardIndexes) (stop bool) {
+		swapFactors = swapFactors.With(poolID, indexes)
+		return false
+	})
+
+	var savingsFactors types.MultiRewardIndexes
+	s.keeper.IterateSavingsRewardIndexes(sdkCtx, func(denom string, indexes types.RewardIndexes) (stop bool) {
+		savingsFactors = savingsFactors.With(denom, indexes)
+		return false
+	})
+
+	var earnFactors types.MultiRewardIndexes
+	s.keeper.IterateEarnRewardIndexes(sdkCtx, func(denom string, indexes types.RewardIndexes) (stop bool) {
+		earnFactors = earnFactors.With(denom, indexes)
+		return false
+	})
+
+	return &types.QueryRewardFactorsResponse{
+		UsdxMintingRewardFactors: usdxFactors,
+		HardSupplyRewardFactors:  supplyFactors,
+		HardBorrowRewardFactors:  borrowFactors,
+		DelegatorRewardFactors:   delegatorFactors,
+		SwapRewardFactors:        swapFactors,
+		SavingsRewardFactors:     savingsFactors,
+		EarnRewardFactors:        earnFactors,
+	}, nil
+}
+
+func (s queryServer) Apy(
+	ctx context.Context,
+	req *types.QueryApyRequest,
+) (*types.QueryApyResponse, error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	params := s.keeper.GetParams(sdkCtx)
+	var apys types.APYs
+
+	// bkava APY (staking + incentive rewards)
+	stakingAPR, err := GetStakingAPR(sdkCtx, s.keeper, params)
+	if err != nil {
+		return nil, err
+	}
+
+	apys = append(apys, types.NewAPY(liquidtypes.DefaultDerivativeDenom, stakingAPR))
+
+	// Incentive only APYs
+	for _, param := range params.EarnRewardPeriods {
+		// Skip bkava as it's calculated earlier with staking rewards
+		if param.CollateralType == liquidtypes.DefaultDerivativeDenom {
+			continue
+		}
+
+		// Value in the vault in the same denom as CollateralType
+		vaultTotalValue, err := s.keeper.earnKeeper.GetVaultTotalValue(sdkCtx, param.CollateralType)
```

Comment:
do we need to do any bounds checking (e.g. greater than 0) on the return value before using it below (or does this method return an error in that case?)

Reply:
yup the `GetAPYFromMultiRewardPeriod` will return 0 for apy if the total value is 0